gforth.1
.TH GForth 1 "November 27, 1994" \" -*- nroff -*-

.SH NAME

GForth

.SH SYNOPSIS

\fCgforth\fR [initialization options] [image-specific options]

.SH DESCRIPTION

\fBGForth\fR is a fast and portable implementation of the Forth
programming language. For details read the manual.

.SH ENVIRONMENT VARIABLES

\fCGFORTHPATH\fR
gives the search path for source files.

.SH EXAMPLES

\fCgforth\fR

starts the system and goes into interactive mode.

\fCgforth file1 file2 -e bye\fR

loads and interprets the files \fCfile1\fR and \fCfile2\fR, then
exits.

.SH SEE ALSO

The GForth manual is available in hypertext (Info, HTML) and printable
(TeX, PS) forms.

The ANS Forth document (i.e., the standard).


NEWS
# User-visible changes between 0.7.0 and 1.0:

In the following words are mentioned that appear the first time in the
documentation; some have been available earlier, but were not
documented.

## Stability

It is now documented which words are intended to be permanently
supported, and which are not; see [Stability
Goals](gforth_html/Stability-Goals.html).

## Headline features

* The text interpreter can recognize things beyond words in the search
  order, integer and FP numbers.

  * Additional ways to write literals in source code: scaled notation
    for FP numbers (`6k5`, `23%`); strings (`"string"`), xts (``
    `word``), nts (``` ``word```), body addresses (`<word> <word+8>`).
    See [Literals in source
    code](gforth_html/Literals-in-source-code.html);

  * You can specify the vocabulary (or several of a word), e.g.,
    `forth:and` vs. `assembler:and`.

  * Nonparsing syntax for dealing with value-flavoured and
    defer-flavoured words (`v` and `d` in the following examples): `->v`
    for `to v`; `+>v` for `+to v`; `'v` for `addr v`; `@>d` for
    `action-of d`; `=>d` for `is d`.  See
    [`rec-to`](gforth_html/Default-recognizers.html#gforth_002dexperimental_002d_002drec_002dto)

  * Contents of environment variables: `${LANG}`

  * You can extend the text interpreter with more **recognizers**:
    `recs rec-name rec-local rec-scope rec-number .-is-dcell?
    rec-float rec-complex rec-string rec-to rec-dtick rec-tick
    rec-body rec-env rec-meta rec-forth rec-sequence: get-recs
    set-recs rec-none translate-name translate-cell translate-dcell
    translate-float translate-complex translate-string
    scan-translate-string translate-env translate-to rec-filter
    rec-forth-nt?  translate: interpreting compiling postponing
    ?rec-found`

* IDE features: Show source code of a word (`locate browse`), show
  uses of a word (`where nw bw ww`); backtraces show source locations
  and can be browsed (`nt bt tt`); and use an editor if desired (`edit
  g gg whereg`); other words: `fancy-after-l xt-locate l n b
  before-locate after-locate ll short-where expand-where prepend-where
  unused-words`

* "Smart `.s`: `...` guesses what stack items mean and displays them
  appropriately. `.s` itself stays dumb.

* A status line at the bottom of the terminal shows stacks, `BASE`,
  stacks, and search order.  `-status +status`

* Multi-Tasking on multiple cores or hardware threads: `newtask task
  newtask4 stacksize stacksize4 initiate execute-task kill-task kill
  stop stop-ns stop-dns thread-deadline halt sleep wake restart pause
  auser uallot up@ user' 's mutex lock unlock critical-section
  atomic!@ atomic+!@ atomic?!@ barrier`

  Message queues (similar to the actor model) provide a way to combine
  communication with synchronization of the tasks: `send-event ?events
  event-loop`

  Cilk is useful for distributing work across tasks: `cores cilk-init
  spawn spawn1 spawn2 cilk-sync cilk-bye`

* C interface enhancements, includig callbacks: `c-value c-variable
  c-funptr c++-library-name c++-library add-libpath add-framework
  add-incdir add-cflags add-ldflags c-callback c-callback-thread
  callback#`

* The dictionary can now have more than one section, each with its own
  dictionary pointer, which allows allocation without affecting other
  sections, used, e.g., for nested definitions.  Words: 'next-section
  previous-section extra-section .sections' See
  [Sections](gforth_html/Sections.html).

* Postpone multiple words with `]]`...`[[`.

* Value-flavoured and defer-flavoured fields: `value: cvalue: wvalue:
    lvalue: scvalue: swvalue: slvalue: 2value: fvalue: sfvalue:
    dfvalue: zvalue: $value: value[]: $value[]: defer:`

* Per-task values and defers: `uvalue udefer`

* Closures: Execution tokens for code with data: `free-closure [n:l
  [d:l [f:l [n:d [d:d [f:d [n:h [d:h [f:h [n:h1 [d:h1 [f:h1 [{: :}l
  :}d :}h :}h1 :}xt >addr <{: ;>`

* New word header layout, and based on that:

  * Modify the behaviour of words: `set-does> set-compsem set->int
    set->comp set-name>string set-name>link`; using a code addres
    instead of an xt: `set-execute`; new code addresses for that:
    `dovalue: dodoes: doabicode:`

  * User-defined `to` etc.: `to-table: n/a to-class: >uvalue set-to`

  * Modify the implementation of words: `set-optimizer`

  * For fans of non-postfix syntax: `compsem: intsem: opt:`

  * Creating a word from a template: `create-from reveal reveal!
    noname-from`

  * Apply `set-does>` etc. to a word defined earlier: `make-latest`

  * Introspection: `.hm immediate?`



## No longer documented words (many of those are still in Gforth 1.0,
  some are not):

  * `.name` (use `id.`);

  * `does-handler!` (implementation changed, use `set-does>`);

  * `/does-handler` (implementation changed, see [Carnal words](gforth_html/Carnal-words.html);

  * `@local# f@local# laddr# lp+!#` (replaced by `@localn f@localn lp+n lp+!`);

  * `clear-libs` (interface changed, use `c-library` ... `end-c-library`)

  * `<interpretation interpretation> <compilation compilation>
    create-interpret/compile` (replaced by `set->int set-compsem`)

  * `naligned` (use `*aligned`)
   
  * `name>comp name>int` (use `name>compile name>interpret`)

  * `path-allot` (us `dup , 0 , allot`)

  * `sl@ sw@ ul@ uw@` (use `l@ l>s`; `w@ w>s`; `l@`; `w@`).

  * Internal and therefore no longer documented: `;s block-position
    cfalign cfaligned compile-lp+! empty-buffer name?int s>number?
    s>unumber? save-buffer`

## Other additional features

* You can use the return stack interpretively (except for passing data
  into or out of INCLUDED, EVALUATE etc.

* `help` shows the documentation of words, or documentation nodes.
  Other words providing information: `authors` `license`.

* Boolean flags: `select`

* Arithmetic words:

  * Single precision: `umin umax`

  * Explicitly floored, symmetric, or unsigned division words: `/s /f
    u/ mods modf umod /mods /modf u/mod du/mod */s */f u*/ */mods

doc/gforth.js
/* filter-search for gforth word index */
/* (c)copyright 2023 by Gerald Wodni  */

function assimilateSearch() {
    const table = document.querySelector("table.index-fn");

    /* insert search box */
    table.insertAdjacentHTML('afterbegin', `
        <tr>
            <td colspan="2">
                <input type="text" id="searchbox" placeholder="filter" autofocus="autofocus"/>
            </td>
        </tr>
    `);
    const searchBox = document.getElementById("searchbox");
    searchBox.focus();

    /* words */
    const sectionTrs = [...table.querySelectorAll("th:first-child")].map( th => th.closest("tr") );
    const emptyTrs = [...table.querySelectorAll("td[colspan='4']")].map( td => td.closest("tr") );
    const wordTrs = [...table.querySelectorAll("code")].map( code => { return {
        word: code.textContent.split(" ")[0],
        tr: code.closest("tr"),
    } });

    /* filter */
    function trVisible( tr, visible ) {
        tr.style.display = visible ? "table-row" : "none";
    }
    searchBox.addEventListener("keyup", evt => {
        const needle = searchBox.value.trim();
        console.log( "SEARCH", needle );
        let viewAll = needle == "";
        wordTrs.forEach( w => trVisible( w.tr, viewAll || w.word.indexOf( needle ) >= 0 ) );
        sectionTrs.forEach( tr => trVisible( tr, viewAll ) );
        emptyTrs.forEach(   tr => trVisible( tr, viewAll ) );
    })
}

if( location.pathname == "/manual/Word-Index.html" )
    assimilateSearch();
else
    console.log( "Unhandles location", location.pathname );
